network/dag/consistency.go

```diff
@@ -0,0 +1,160 @@
+	"github.com/nuts-foundation/nuts-node/network/dag/tree"
+	"github.com/nuts-foundation/nuts-node/network/log"
+	"sync"
+	"time"
+)
+
+type circuit int
+
+const (
+	circuitGreen circuit = iota
+	circuitYellow
+	circuitRed
+)
+
+// xorTreeRepair is responsible for repairing the XOR tree. Its loop is triggered when the network layer detects differences in XOR values with all other nodes.
+// It will loop over all pages of the XOR tree and recalculates the XOR value with the transactions in the database.
+// This repair is needed because current networks have nodes that have a wrong XOR value. How this happens is not yet known, it could be due to DB failures of due to failures in older versions.
+// The fact is that we can fix the state relatively easy.
+// The loop checks a page (512 LC values) per 10 seconds and continues looping until the network layer signals all is ok again.
+type xorTreeRepair struct {
+	ctx          context.Context
+	cancel       context.CancelFunc
+	ticker       *time.Ticker
+	currentPage  uint32
+	state        *state
+	circuitState circuit
+	mutex        sync.Mutex
+}
+
+func newXorTreeRepair(state *state) *xorTreeRepair {
+	return &xorTreeRepair{
+		state:  state,
+		ticker: time.NewTicker(10 * time.Second),
+	}
+}
+
+func (f *xorTreeRepair) start() {
+	var ctx context.Context
+	ctx, f.cancel = context.WithCancel(context.Background())
+	go func() {
+		for {
+			select {
+			case <-ctx.Done():
+				return
+			case <-f.ticker.C:
+				f.checkPage()
+			}
+		}
+	}()
+}
+
+func (f *xorTreeRepair) shutdown() {
+	if f.cancel != nil {
+		f.cancel()
+	}
+}
+
+func (f *xorTreeRepair) checkPage() {
+	f.mutex.Lock()
+	defer f.mutex.Unlock()
+
+	// ignore run if circuit is not red
+	if f.circuitState < circuitRed {
+		return
+	}
+
+	currentLC := f.state.lamportClockHigh.Load()
+	lcStart := f.currentPage * PageSize
+	lcEnd := lcStart + PageSize
+
+	// initialize an XOR tree
+	calculatedXorTree := tree.New(tree.NewXor(), PageSize)
+
+	// acquire global lock
+	err := f.state.graph.db.Write(context.Background(), func(txn stoabs.WriteTx) error {
+		txs, err := f.state.graph.findBetweenLC(txn, lcStart, lcEnd)
+		if err != nil {
+			return err
+		}
+		for _, tx := range txs {
+			calculatedXorTree.Insert(tx.Ref(), tx.Clock())
+		}
+
+		// Get XOR leaf from current XOR tree
+		xorTillEnd, _ := f.state.xorTree.getZeroTo(lcEnd - 1)
+		if lcStart != 0 {
+			xorTillStart, _ := f.state.xorTree.getZeroTo(lcStart - 1)
+			_ = xorTillEnd.Subtract(xorTillStart)
+		}
+
+		// Subtract the calculated tree, should be empty if the trees are equal
+		_ = xorTillEnd.Subtract(calculatedXorTree.Root())
+		if !xorTillEnd.Empty() {
+			// it's not empty, so replace the leaf in the current XOR tree with the calculated one
+			err = f.state.xorTree.tree.Replace(lcStart, calculatedXorTree.Root())
+			if err != nil {
+				return err
+			}
+			log.Logger().Warnf("detected XOR tree mismatch for page %d, fixed using recalculated values", f.currentPage)
```

Comment:
could it just be it's temporarily out-of-sync with the other nodes due "normal" out-of-date (missing a few TXs) and being slow for a moment?

Reply:
only when you also have experienced down time. In a high traffic scenario, it could be triggered once. The checker starts at 0 so it'll start checking the page that has no traffic to avoid racy fixes.